src/merkle_tree.rs

```diff
@@ -0,0 +1,221 @@
+use hmac_sha256::Hash;
+
+pub struct MerkleTree {
+    pub root: Option<[u8; 32]>,
+    pub leafs: Vec<[u8; 32]>,
+}
+
+impl Default for MerkleTree {
+    fn default() -> Self {
+        Self {
+            root: None,
+            leafs: Vec::new(),
+        }
+    }
+}
+
+// TODO: type Hash for [u8; 32]
+// TODO: type Level for Vec<[u8; 32]> of even length (duplicates last element if odd)
+
+impl MerkleTree {
+    /// Create a new empty MerkleTree instance.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Create a new MerkleTree from the provided items.
+    /// Each item should be representable as bytes.
+    /// It returns a `MerkleTree` instance with the leaf hashes and the Merkle root.
+    pub fn build<T: AsRef<[u8]>>(&self, items: &[T]) -> Self {
+        let leafs: Vec<[u8; 32]> = items.iter().map(|item| self.hash(item.as_ref())).collect();
+
+        Self {
+            root: Some(self.merkle_root(leafs.clone())),
+            leafs,
+        }
+    }
```

Comment:
This function should probably check that ´items´ is not empty.

Reply:
So true!

Do you think it should `Err` if there were no items? Or just return an empty tree?

Reply:
Given that the tree will have the feature of dynamically adding elements I think the second option would be better.